README.md
# Blockchain Data

Blockchain has the potential to change the way that the world approaches data. Develop Blockchain skills by understanding the data model behind Blockchain by developing your own simplified private blockchain.

## Getting Started

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes.

### Prerequisites

Installing Node and NPM is pretty straightforward using the installer package available from the (Node.js® web site)[https://nodejs.org/en/].

### Configuring your project

- Use NPM to install all the project dependencies.

```javascript
npm install
```

- This project depends on following packages

    - [`crypto-js`](https://www.npmjs.com/package/crypto-js) - Code uses SHA256 hashing from this package
    - [`level db`](https://github.com/Level/levelup) - to provide a key/value persistence store
    - [`express`](https://expressjs.com/) - to provide middleware for writing api server
    - [`bitcoinjs-lib`](https://www.npmjs.com/package/bitcoinjs-lib) - A javascript Bitcoin library for node.js and browsers.
    - [`bitcoinjs-message`](https://www.npmjs.com/package/bitcoinjs-message) - A javascript library to sing/verify message signatures using `bitcoinjs-lib`
    - [`hex2ascii`](https://www.npmjs.com/package/hex2ascii) - Convert hex to ascii in JavaScript. 

## Running the code

- run api server with:

```javascript
node app.js
```

This will start an api server at http://localhost/8000

The server right now supports six api endpoints as explained below.

### Endpoint documentation

#### 1. POST VALIDATION REQUEST

End point is to generate a request for validation with payload containing a address.

##### URL

"post" to

```javascript
http://localhost:8000/requestValidation
```

with request body containing 

```javascript
{
    "address":"12m14JXqX6DbwwFoScrUyaSPhPTbheuGfT"
}
```

##### RESPONSE

Registers a request in mempool and returns the request Object

```javascript
{
    "address": "12m14JXqX6DbwwFoScrUyaSPhPTbheuGfT",
    "requestTimeStamp": "1542453128",
    "message": "12m14JXqX6DbwwFoScrUyaSPhPTbheuGfT:1542453128:starRegistry",
    "validationWindow": 300
}
```

`validationWindow` is the time in seconds this request will be valid and stay in mempool.

#### 2. POST VERIFY VALIDATION REQUEST

End point is to verify a request for validation with payload containing a address and message signature.

You can use the `message` returned by above request and sign it with the `address` using [`Electrum`](https://electrum.org/#home). 

##### URL

"post" to

```javascript
http://localhost:8000/message-signature/validate
```

with request body containing 

```javascript
{
    "address": "12m14JXqX6DbwwFoScrUyaSPhPTbheuGfT",
    "signature": "HztubDBR2CzMneybxoBXcukJhTNmQMMzX46qtnVE6IwqKLbAtuWSxEhingrfi+nx8/wrQC3YzKNbqY0E8GqjMzk="
}
```

##### RESPONSE

Validates a request registered in mempool and returns the valid request Object

```javascript
{
    "registerStar": true,
    "status": {
        "address": "12m14JXqX6DbwwFoScrUyaSPhPTbheuGfT",
        "requestTimeStamp": "1542453128",
        "message": "12m14JXqX6DbwwFoScrUyaSPhPTbheuGfT:1542453128:starRegistry",
        "validationWindow": 1800,
        "messageSignature": "valid"
    }
}
```

#### 3. POST STAR DATA

End point is to register a star data in blockchain for a valid request.

##### URL

"post" to

```javascript
http://localhost:8000/block
```

with request body containing address and star data.

```javascript
{
    "address": "12m14JXqX6DbwwFoScrUyaSPhPTbheuGfT",
    "star": {
        "dec": "68° 52' 53.9",
        "ra": "16h 29m 1.0s",
        "story": "Found star using telescope"
    }
}
```

##### RESPONSE

Registers a star against the address in blockchain. Returns the newly added block.

```javascript
{
    "hash": "0ffb3bdeaf779fb297d357b1e3b79e1f7f15337e3ca91aa715b3279cdf3f4df6",
    "height": 2,
    "body": {
        "address": "12m14JXqX6DbwwFoScrUyaSPhPTbheuGfT",
        "star": {
            "dec": "68° 52' 53.9",
            "ra": "16h 29m 1.0s",
            "story": "466f756e642073746172207573696e672074656c6573636f7065",
            "storyDecoded": "Found star using telescope"
        }
    },
    "time": "1542453200",
    "previousBlockHash": "422e46d1b86fa534d5054496a5efc8d9152eebc0d38a74db52845faf9f30561c"
}
```

#### 4. GET BLOCK BY HASH

End point is to retrieve a block by hash value.

##### URL
"get" to

```javascript
http://localhost:8000/stars/hash:<hasvalue>
```

e.g.

```javascript
http://localhost:8000/stars/hash:422e46d1b86fa534d5054496a5efc8d9152eebc0d38a74db52845faf9f30561c
```

##### RESPONSE

Returns a block from blockchain matching the hash value.

```javascript
{
    "hash": "422e46d1b86fa534d5054496a5efc8d9152eebc0d38a74db52845faf9f30561c",
    "height": 1,
    "body": {
        "address": "12m14JXqX6DbwwFoScrUyaSPhPTbheuGfT",
        "star": {
            "dec": "68° 52' 56.9",
            "ra": "16h 29m 1.0s",
            "story": "466f756e642073746172207573696e672068747470733a2f2f7777772e676f6f676c652e636f6d2f736b792f",
            "storyDecoded": "Found star using https://www.google.com/sky/"
        }
    },
    "time": "1542449148",
    "previousBlockHash": "18053a73a6597e7a73e50dc47829ad0351bce84748956f2bb61840c9edd22fb9"
}
```

#### 5. GET BLOCK BY BLOCK HEIGHT

End point is to retrieve a block by blockheight.

##### URL

"get" to

```javascript
http://localhost:8000/block/<blockheight>
```

e.g.

```javascript
http://localhost:8000/block/1
```

##### RESPONSE

Returns a block from blockchain matching blockheight.

```javascript
{
    "hash": "422e46d1b86fa534d5054496a5efc8d9152eebc0d38a74db52845faf9f30561c",
    "height": 1,
    "body": {
        "address": "12m14JXqX6DbwwFoScrUyaSPhPTbheuGfT",
        "star": {
            "dec": "68° 52' 56.9",
            "ra": "16h 29m 1.0s",
            "story": "466f756e642073746172207573696e672068747470733a2f2f7777772e676f6f676c652e636f6d2f736b792f",
            "storyDecoded": "Found star using https://www.google.com/sky/"
        }
    },
    "time": "1542449148",
    "previousBlockHash": "18053a73a6597e7a73e50dc47829ad0351bce84748956f2bb61840c9edd22fb9"
}
```

#### 6. GET BLOCK(S) BY WALLET ADDRESS

End point is to retrieve block(s) by wallet address.

##### URL

"get" to

```javascript
http://localhost:8000/stars/address:<address>
```

e.g.

```javascript
http://localhost:8000/stars/address:12m14JXqX6DbwwFoScrUyaSPhPTbheuGfT
```

##### RESPONSE

Returns block(s) from blockchain matching wallet address.

```javascript
[
    {
        "hash": "422e46d1b86fa534d5054496a5efc8d9152eebc0d38a74db52845faf9f30561c",
        "height": 1,
        "body": {
            "address": "12m14JXqX6DbwwFoScrUyaSPhPTbheuGfT",
            "star": {
                "dec": "68° 52' 56.9",
                "ra": "16h 29m 1.0s",
                "story": "466f756e642073746172207573696e672068747470733a2f2f7777772e676f6f676c652e636f6d2f736b792f",
                "decodedStory": "Found star using https://www.google.com/sky/"
            }
        },
        "time": "1542449148",
        "previousBlockHash": "18053a73a6597e7a73e50dc47829ad0351bce84748956f2bb61840c9edd22fb9"
    },
    {
        "hash": "0ffb3bdeaf779fb297d357b1e3b79e1f7f15337e3ca91aa715b3279cdf3f4df6",
        "height": 2,
        "body": {
            "address": "12m14JXqX6DbwwFoScrUyaSPhPTbheuGfT",
            "star": {
                "dec": "68° 52' 53.9",
                "ra": "16h 29m 1.0s",
                "story": "466f756e642073746172207573696e672074656c6573636f7065",
                "decodedStory": "Found star using telescope"
            }
        },
        "time": "1542453200",
        "previousBlockHash": "422e46d1b86fa534d5054496a5efc8d9152eebc0d38a74db52845faf9f30561c"
    }
]
```

## Code Explanation

In this project a simple private blockchain is implemented with functionality to create new blocks and add them to block chain. The data is persisted in `leveldb` database. Additional functionality includes ability to validate individual blocks or the entire blockchain. 

- `Block.js` and `Blockchain.js` contain implementation of blockchain.
- `Mempool.js` contains implementation of a simple mempool to register and validate requests, holding these in memory till timeouts.
- `app.js` contains the code for `api` server. The end points are explained above.
- `testBlockchain.js` contains some basic test cases to test bloclchain code as explained below.

## Testing

To test code:

1: Open a command prompt or shell terminal after installing node.js.

2: Run test cases in file `testBlockchain.js` using following command:

```javascript
node testBlockchain.js
```

This will check implementation of blockchain code in `Blockchain.js`. 

Testing sequence is:

- create Blockchain instance
- Insert 10 new blocks into the chain
- print the data in entire blockchain
- validate the chain
- modify data(tamper) in one of the blocks
- revalidate chain and see validation failing for modified block 


changelog.txt
1. http://localhost:8000/requestValidation
Comment from reviewer: 
"Response has correct format – field walletAddress should be renamed to address"
My response: 
Though I have renamed the field, project instructions are inconsistent. 
Some places they have used "walletAddress" and some places "address". 
I do not think exact name of the field is really that important esp given 
the inconsistency in instructions. Please check field names in 
point 4 (requestObject)in "Project Walk-through" 
vs point 6 (validationRequestByWallet)


2. http://localhost:8000/message-signature/validate
Comment from reviewer:
Response has correct format – field walletAddress should be renamed to address
Also validationWindow field should be less than maximum value (300) 
My Response:
"walletAddress" vs "address" - same comment as above.
ValidationWindow - my code is consistent with revised project instructions of 1800 sec timeout 
for valid requests and 300 sec for validation requests

3. Reviewer Comments + Specs: Star story supports ASCII text, limited to 250 words (500 bytes), and hex encoded.
My comment: How is one word always 2 bytes? Or was the spec meant to be 250 characters? 
I have modified to strip out non ASCII characters and restrict the comments to first 250 characters

4. Reviewer: "AddBlock" should not contain "storyDecoded". 
My comment. Please check point 11 in Project walk-through which shows storyDecoded to be 
part of response (though not stored in block)



Mempool.js
const bitcoinMessage = require('bitcoinjs-message'); 


const TimeoutRequestsWindowTime = 5*60*1000;
const TimeoutMempoolValidWindowTime = 30*60*1000;

class Mempool {

  constructor() {
    this.mempool = [];
    this.timeoutRequests = [];
    this.mempoolValid = [];
    this.timeoutMempoolValid = [];
  }


  // method to add a new validation request or to retrieve an existing one
  addRequestValidation(address) {

    if (address in this.mempool) {
      const req = this.mempool[address];
      let timeElapse = (new Date().getTime().toString().slice(0,-3)) - req.requestTimeStamp;
      let timeLeft = (TimeoutRequestsWindowTime/1000) - timeElapse;
      req.validationWindow = timeLeft;
      return req;
    } else {
      let timeStamp = (new Date().getTime().toString().slice(0,-3));
      const req = {
        address: address,
        requestTimeStamp: timeStamp,
        message:`${address}:${timeStamp}:starRegistry`,
        validationWindow: TimeoutRequestsWindowTime/1000
      }

      // add request to mempool
      this.mempool[address] = req;
      // set timout to remove entry from mempool
      this.timeoutRequests[req.address] = 
        setTimeout(() => this.removeRequestValidation(req.address),
                    TimeoutRequestsWindowTime)
      return req;
    }
  }

  // method to remove validationrequest on timeout
  removeRequestValidation(address) {
    //filterout existing entry with address from mempool
    this.mempool = this.mempool.filter(elem => elem.address !== address);
  }


  //method to validates request by address
  validateRequestByWallet(address, signature) {

    if (address in this.mempoolValid) {
      if (bitcoinMessage.verify(this.mempoolValid[address].status.message, address, signature)) {
        const req = this.mempoolValid[address];
        console.log(req);
        let timeElapse = (new Date().getTime().toString().slice(0,-3)) - req.status.requestTimeStamp;
        let timeLeft = (TimeoutMempoolValidWindowTime/1000) - timeElapse;
        req.status.validationWindow = timeLeft;
        return req;
      } else {
        return {message:'Signature invlaid. Please try again.'};
      }
    }

    if (address in this.mempool) {
      if (bitcoinMessage.verify(this.mempool[address].message, address, signature)) {
        const req = {
          registerStar: true,
          status: Object.assign({}, 
                      this.mempool[address], 
                      { validationWindow: TimeoutMempoolValidWindowTime/1000,
                        messageSignature: 'valid',
                      })
          }

        clearTimeout(this.timeoutRequests[address]); //clear timout
        this.removeRequestValidation(address); // remove request from mempool
        // add request to valid request mempool
        this.mempoolValid[address] = req;
        // set timout to remove entry from valid rquest mempool
        this.timeoutMempoolValid[address] = 
          setTimeout( () => this.removeRequestByWallet(address),
                      TimeoutMempoolValidWindowTime)
        return req;
      } else {
        return {message:'Signature invlaid. Please try again.'};
      }
    } else {
      return { message: "Validation request expired or not created." };
    }
  }

  // method to remove validation request on timeout
  removeRequestByWallet(address) {
    //filterout existing entry with address from mempool
    this.mempoolValid = this.mempoolValid.filter(elem => elem.address !== address);
  }

  // verify if there is a valid request in mempoolValid
  verifyAddressRequest(address) {
    return address in this.mempoolValid;
  }

  // remove request and clear SetTimeout
  removeAddressRequest(address) {
    clearTimeout(this.timeoutMempoolValid[address]); //clear timout
    this.removeRequestByWallet(address); // remove request from mempool
  }

}

module.exports = Mempool;

Blockchain.js
/* ===== SHA256 with Crypto-js ===============================
|  Learn more: Crypto-js: https://github.com/brix/crypto-js  |
|  =========================================================*/

const SHA256 = require('crypto-js/sha256');

/* ===== Persist data with LevelDB ===========================
|  Learn more: level: https://github.com/Level/level         |
|  =========================================================*/

const level = require('level');
const chainDB = './chaindata';
const db = level(chainDB);


/*======= Import Block class =================================
|  Block class                                               |
|===========================================================*/
const Block = require('./Block');

/* ===== Blockchain Class ===========================
|  Class with a constructor for new blockchain 		  |
|  ================================================*/

class Blockchain {

constructor() {
    return (async () => {
      try {
        const height = await this.getBlockHeight();
      // if no genesis block, add that
        if (height === 0) {
          await this.addGenesisBlock();
        }
      } catch(err) {
        console.log("Error in init: "+ err);
      }
      return this;
    })();
  }

  async addGenesisBlock() {
    const genesis_block = new Block("First block in the chain - Genesis block");
    genesis_block.height = 0;
    genesis_block.time = new Date().getTime().toString().slice(0,-3);
    genesis_block.hash = await SHA256(JSON.stringify(genesis_block)).toString();
    // Adding genesis block object to chain
    try {
      await db.put(genesis_block.height, JSON.stringify(genesis_block).toString());
      console.log("Genesis block added");
    } catch(err) {
      console.log("error inside addGenesisBlock:" + err);
      throw err;
    }

  }


  // Add new block
  async addBlock(newBlock) {

    // Block height
    try {
      const height = await this.getBlockHeight();
      //console.log("addBlock:Height="+height);
    // if no genesis block, first add that
    if (height === 0) {
      await this.addGenesisBlock();
      height ++;
    }
    newBlock.height = height;

    // UTC timestamp
    newBlock.time = new Date().getTime().toString().slice(0,-3);

    // previous block hash
    let prevBlock = await this.getBlock(newBlock.height-1);
    newBlock.previousBlockHash = prevBlock.hash;

    // Block hash with SHA256 using newBlock and converting to a string
    newBlock.hash = await SHA256(JSON.stringify(newBlock)).toString();
    // Adding block object to chain
    await db.put(newBlock.height, JSON.stringify(newBlock).toString());
    return newBlock;
    } catch (err) {
      console.log("Error in AddBlock:"+err);
      throw err;
    }
  }

  // Get block height
  getBlockHeight() {
    return new Promise((resolve, reject) => {
      let i = 0;
      db.createReadStream().on('data', (data) => {
            i++;
          }).on('error', (err) => {
              reject(err);
          }).on('close', () => {
            //console.log('Block Height' + i);
            resolve(i);
          });
    });
  } 
    

  // get block
  async getBlock(blockHeight) {
    try {
      //console.log("fetching block from db with id: "+ blockHeight);
      let blockVal = await db.get(blockHeight);
      return JSON.parse(blockVal); 
    } catch(err) {
      console.log("Error in getBlock:"+err);
      throw err;
    }
  }

  // get blcok by hash
  async getBlockByHash(blockHash) {
    return new Promise((resolve, reject) => {
      let block = null;
      db.createReadStream()
      .on('data', (data) => {
          const value = JSON.parse(data.value);
          if (value.hash === blockHash) {
            block = value;
          }
      })
      .on('error', (err) => {
          reject(err);
      })
      .on('close', () => {
            resolve(block);
      });
    });
  }

  // get block(s) by address
  async getBlocksByAddress(address) {
    return new Promise((resolve, reject) => {
      let blocks = [];
      db.createReadStream()
      .on('data', (data) => {
          const value = JSON.parse(data.value);
          if (value.body.address === address) {
            blocks.push(value);
          }
      })
      .on('error', (err) => {
          reject(err);
      })
      .on('close', () => {
            resolve(blocks);
      });
    });
  }


  // put block
  async putBlock(key, block) {
    try {
      const blockValue = JSON.stringify(block).toString();
      let result = await db.put(key, blockValue);
    } catch(err) {
      console.log("Error in putBlock:"+err);
    }
  }

  // validate block
  async validateBlock(blockHeight){
    // get block object
    let block = await this.getBlock(blockHeight);
    // get block hash
    let blockHash = block.hash;
    // remove block hash to test block integrity
    block.hash = '';
    // generate block hash
    let validBlockHash = await SHA256(JSON.stringify(block)).toString();
    // Compare
    if (blockHash===validBlockHash) {
      return true;
    } else {
      console.log('Block #'+blockHeight+' invalid hash:\n'+blockHash+'<>'+validBlockHash);
      return false;
    }
  }

  // Validate blockchain
  async validateChain(){
    let errorLog = [];
    const height = await this.getBlockHeight();
    let blockHash = '';
    let previousHash = '';
    for (var i = 0; i < height; i++) {
      // validate block
      if (!await this.validateBlock(i))errorLog.push(i);

      // get block with key=i
      let block = await this.getBlock(i);
      previousHash = block.previousBlockHash;
      if (blockHash!==previousHash) {
        // compare blocks hash link
        // except for genesis block
          errorLog.push(i);
      }
      blockHash = block.hash;
    }

    if (errorLog.length>0) {
      console.log('Block errors = ' + errorLog.length);
      console.log('Blocks: '+errorLog);
    } else {
      console.log('No errors detected');
    }
  }
}

module.exports = Blockchain;

simpleChain.js
/* ===== SHA256 with Crypto-js ===============================
|  Learn more: Crypto-js: https://github.com/brix/crypto-js  |
|  =========================================================*/

const SHA256 = require('crypto-js/sha256');

/* ===== Persist data with LevelDB ===================================
|  Learn more: level: https://github.com/Level/level     |
|  =============================================================*/

const level = require('level');
const chainDB = './chaindata';
const db = level(chainDB);



/* ===== Block Class ==============================
|  Class with a constructor for block 			   |
|  ===============================================*/

class Block{
	constructor(data){
     this.hash = "",
     this.height = 0,
     this.body = data,
     this.time = 0,
     this.previousBlockHash = ""
    }
}

/* ===== Blockchain Class ==========================
|  Class with a constructor for new blockchain 		|
|  ================================================*/

class Blockchain {

constructor() {
    return (async () => {
      try {
        const height = await this.getBlockHeight();
      // if no genesis block, add that
        if (height === 0) {
          await this.addGenesisBlock();
        }
      } catch(err) {
        console.log("Error in init: "+ err);
      }
      return this;
    })();
  }

  async addGenesisBlock() {
    const genesis_block = new Block("First block in the chain - Genesis block");
    genesis_block.height = 0;
    genesis_block.time = new Date().getTime().toString().slice(0,-3);
    genesis_block.hash = await SHA256(JSON.stringify(genesis_block)).toString();
    // Adding genesis block object to chain
    try {
      await db.put(genesis_block.height, JSON.stringify(genesis_block).toString());
      console.log("Genesis block added");
    } catch(err) {
      console.log("error inside addGenesisBlock:" + err);
      throw err;
    }

  }


  // Add new block
  async addBlock(newBlock) {

    // Block height
    try {
      const height = await this.getBlockHeight();
      //console.log("addBlock:Height="+height);
    // if no genesis block, first add that
    if (height === 0) {
      await this.addGenesisBlock();
      height ++;
    }
    newBlock.height = height;

    // UTC timestamp
    newBlock.time = new Date().getTime().toString().slice(0,-3);

    // previous block hash
    let prevBlock = await this.getBlock(newBlock.height-1);
    newBlock.previousBlockHash = prevBlock.hash;

    // Block hash with SHA256 using newBlock and converting to a string
    newBlock.hash = await SHA256(JSON.stringify(newBlock)).toString();
    // Adding block object to chain
    await db.put(newBlock.height, JSON.stringify(newBlock).toString());
    return newBlock;
    } catch (err) {
      console.log("Error in AddBlock:"+err);
      throw err;
    }
  }

  // Get block height
  getBlockHeight() {
    return new Promise(resolve => {
      let i = 0;
      db.createReadStream().on('data', (data) => {
            i++;
          }).on('error', (err) => {
              return console.log('Unable to read data stream!', err)
          }).on('close', () => {
            //console.log('Block Height' + i);
            resolve(i);
          });
    });
  } 
    

  // get block
  async getBlock(blockHeight) {
    try {
      console.log("fetching block from db with id: "+ blockHeight);
      let blockVal = await db.get(blockHeight);
      return JSON.parse(blockVal); 
    } catch(err) {
      console.log("Error in getBlock:"+err);
      throw err;
    }
  }

  // put block
  async putBlock(key, block) {
    try {
      const blockValue = JSON.stringify(block).toString();
      let result = await db.put(key, blockValue);
    } catch(err) {
      console.log("Error in putBlock:"+err);
    }
  }

  // validate block
  async validateBlock(blockHeight){
    // get block object
    let block = await this.getBlock(blockHeight);
    // get block hash
    let blockHash = block.hash;
    // remove block hash to test block integrity
    block.hash = '';
    // generate block hash
    let validBlockHash = await SHA256(JSON.stringify(block)).toString();
    // Compare
    if (blockHash===validBlockHash) {
      return true;
    } else {
      console.log('Block #'+blockHeight+' invalid hash:\n'+blockHash+'<>'+validBlockHash);
      return false;
    }
  }

  // Validate blockchain
  async validateChain(){
    let errorLog = [];
    const height = await this.getBlockHeight();
    let blockHash = '';
    let previousHash = '';
    for (var i = 0; i < height; i++) {
      // validate block
      if (!await this.validateBlock(i))errorLog.push(i);

      // get block with key=i
      let block = await this.getBlock(i);
      previousHash = block.previousBlockHash;
      if (blockHash!==previousHash) {
        // compare blocks hash link
        // except for genesis block
          errorLog.push(i);
      }
      blockHash = block.hash;
    }

    if (errorLog.length>0) {
      console.log('Block errors = ' + errorLog.length);
      console.log('Blocks: '+errorLog);
    } else {
      console.log('No errors detected');
    }
  }
}

module.exports = {
 Blockchain: Blockchain,
 Block: Block
}

testBlockchain.js
const Block = require('./Block');
const Blockchain = require('./Blockchain');


//add "i" blocks to chain
async function createChain(chain, n) {
  let newBlock = new Block("Test Block - " + (n + 1));
  await chain.addBlock(newBlock);
  if (n < 10) {
    await createChain(chain, n+1);
  }
}


//print chain
async function printChain(chain, n) {
  let val = await chain.getBlock(n);
  console.log(val);
  if (n < 10) {
    await printChain(chain, n+1);
  }
}

async function modifyBlock(chain) {
  const key = 1;
  let block = await chain.getBlock(key);
  block.body = "Modified Data";
  await chain.putBlock(key, block);
}
/*=====  Testing the functionality =========================|
| add 10 new blcoks to chain                                |
| print the chain value                                     |
| validate chain                                            |
| modify data in block 1                                    |
| revalidate chain                                          |
|+=====  Testing the functionality ========================*/


setTimeout(async () => {
  //init block chain with genesis block
  let blockchain = await new Blockchain();
  await createChain(blockchain, 0);
  console.log('');
  console.log('Block chain created');
  console.log('');
  console.log('Printing block chain data...');
  await printChain(blockchain, 0);
  console.log('');
  console.log('Validating blockchain....');
  await blockchain.validateChain();
  console.log('');
  console.log('Modifying data in block 1...');
  console.log('data before modification...');
  console.log(await blockchain.getBlock(1));
  console.log('');
  await modifyBlock(blockchain);
  console.log("data after modification...");
  console.log(await blockchain.getBlock(1));
  console.log('');
  console.log('Validating blockchain after modification...')
  await blockchain.validateChain();
  console.log('');
  console.log('Creating a new instance of blcokchain...')
  let blockchain_2 = await new Blockchain();
  console.log('Printing block chain - only one Gneesis block is created')
  await printChain(blockchain_2, 0);
}, 1000);


app.js
const express = require('express');
const app = express();
const bodyParser = require('body-parser');

//import blockchain classes from simplechain.js
const Block = require('./Block');
const Blockchain = require('./Blockchain');

// import Mempool class
const Mempool = require('./Mempool');

//import hex2ascii module
const hex2ascii = require('hex2ascii');

// configure app to use bodyParser()
// this will allow us to get data from a POST
app.use(bodyParser.urlencoded({ extended: true }));
app.use(bodyParser.json());

// set port
const port = 8000;


// variable holding blockchain instance
// will be initialised in a async function at the time of 
// starting the server at the end of this file
let blockchain = null;

// variable holding mempool
let mempool = new Mempool();

// Routes for our api
// ====================================================
const router = express.Router();

router.get('/', function(req, res) {
  res.json({ message: 'Welcome to blockchain api!.' });
});


// ==============================================================
// 1st route is a post to http://localhost:8000/requestValidation
// This endpoint validates the request
// ==============================================================

router.route('/requestValidation').post((req, res) => {
  const address = req.body.address;

  // send 400 if body does not contain wallet address
  if (!address) {
    res.status(400).json({message: "Request body does not contain wallet address i.e. a field named 'address'."});
  } else {
    res.json(mempool.addRequestValidation(address));
  }

});

// ==============================================================
// 2nd route is a post to http://localhost:8000/message-signature/validate
// This endpoint validates the signed request
// ==============================================================
router.route('/message-signature/validate').post((req, res) => {
  const address = req.body.address;
  const signature = req.body.signature;
  if (!address || !signature) {
    res.status(400).json({message: "Request body requires a address and signature."});
  } else {
    res.json(mempool.validateRequestByWallet(address, signature));
  }
});


// ==============================================================
// 3rd route is a post to http://localhost:8000/block
// This endpoint registers a star request 
// and returns the added block in blockchain
// ==============================================================
router.route('/block').post(async (req, res) => {
  const address = req.body.address;
  const star = req.body.star;
  if (!address || !star) {
    res.status(400).json({message: "Request body requires a address and a star to be registered."});
  } else if (!star.dec || !star.ra || !star.story) {
    res.status(400).json({message: "'star' object must contain coordinates 'ra' & 'dec' as well as 'story'."});
  } else if (!mempool.verifyAddressRequest(address)) {
  // check if there is a verified request in mempool
    res.json({ message: "No verified request found" });
  } else {
    // add new block to chain
    let storyDecoded = star.story.replace(/[^A-Za-z 0-9 \.,\?""!@#\$%\^&\*\(\)-_=\+;:<>\/\\\|\}\{\[\]`~]*/g, '');
    // accept only 250 characters
    storyDecoded = storyDecoded.substring(0,250);
    const storyEncoded = Buffer(storyDecoded).toString('hex');
    const blockBody = {address,
                       star}
    blockBody.star.story = storyEncoded;
    console.log('blockBody:'+ blockBody);
    const block = new Block(blockBody);
    try {
      const newBlock = await blockchain.addBlock(block);
      newBlock.body.star.storyDecoded = hex2ascii(newBlock.body.star.story);
      // remove the valid request from mempool
      mempool.removeAddressRequest(address);
      console.log("Added new Block:");
      console.log(newBlock);
      res.json(newBlock);
    } catch (err) {
      console.log("Error in post block:" + err);
      res.status(500).send("something broke in server");
    }
  }
});


// ==============================================================
// 4th route is a post to http://localhost:8000/stars/hash:<hashvalue>
// This endpoint returns a block with matching hashvalue 
// ==============================================================
router.route('/stars/hash::hash').get(async function(req, res) {
  console.log("need to fetch block with hash: " + req.params.hash);
  try {
    let block = await blockchain.getBlockByHash(req.params.hash);
    if (block) {
      block.body.star.storyDecoded = hex2ascii(block.body.star.story);
      res.json(block);
    } else {
      res.json({ message: 'block not found.'})
    }
  } catch (err) {
    //console.log(err);
    if (err.notFound) {
      res.status(404).send('block hash: "' + req.params.hash + '" not found' );
    } else {
      console.log("Error in getBlockByHash: "+ err);
      res.status(500).send("something broke in server");
    }
  }
});

// ====================================================
// 5th route is a get to http://localhost:8000/stars/address:<address>
// return all the blocks with given address
// ====================================================
router.route('/stars/address::address').get(async function(req, res) {
  console.log("need to fetch blocks with address: " + req.params.address);
  try {
    let blocks = await blockchain.getBlocksByAddress(req.params.address);
    if (blocks) {
      blocks.forEach(elem => {
        elem.body.star.decodedStory = hex2ascii(elem.body.star.story); 
      });
      res.json(blocks);
    } else {
      res.json({ message: 'blocks not found.'})
    }
  } catch (err) {
    //console.log(err);
    if (err.notFound) {
      res.status(404).send('block hash: "' + req.params.hash + '" not found' );
    } else {
      console.log("Error in getBlockByHash: "+ err);
      res.status(500).send("something broke in server");
    }
  }
});


// ====================================================
// 6th route is a get to http://localhost:8000/block/:id
// return the block with blockheight = id
// ====================================================
router.route('/block/:blockheight').get(async function(req, res) {
  console.log("need to fetch block: " + req.params.blockheight);
  try {
    let block = await blockchain.getBlock(req.params.blockheight);
    block.body.star.storyDecoded = hex2ascii(block.body.star.story);
    res.json(block);
  } catch (err) {
    //console.log(err);
    if (err.notFound) {
      res.status(404).send('block: "' + req.params.blockheight + '" not found' );
    } else {
      console.log("Error in getBlock: "+ err);
      res.status(500).send("something broke in server");
    }
  }
});

// configure api end point to base URL
app.use('/', router);


// Start Server
// ====================================================
(async () => {
  //start blockchain instance
  blockchain = await new Blockchain();
  // start api server and listen to requests at port=8000
  app.listen(port);
  console.log('Blockchain api server running at post ' + port);
})();
